certutils/reloader.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This code is based on the example in https://github.com/kelseyhightower/kube-cert-manager/blob/master/tls-app/certificate-manager.go
+
+package certutils
+
+import (
+	"crypto/tls"
+	"log"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// CertReloader provides a mechanism for reloading a TLS key and cert upon file change
+type CertReloader struct {
+	sync.RWMutex
+	certificate *tls.Certificate
+	certFile    string
+	keyFile     string
+	Error       chan error
+	watcher     *fsnotify.Watcher
+}
+
+// NewCertReloader returns a new CertReloader
+func NewCertReloader(certFile, keyFile string) (*CertReloader, error) {
+	cr := &CertReloader{
+		certFile: certFile,
+		keyFile:  keyFile,
+		Error:    make(chan error, 10),
+	}
+	err := cr.setCertificate()
+	if err != nil {
+		return nil, err
+	}
+
+	go cr.watchCertificate()
+
+	return cr, nil
+}
+
+// GetCertificate implements the tls.Config GetCertificate() func
+func (cr *CertReloader) GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	cr.RLock()
+	defer cr.RUnlock()
+	return cr.certificate, nil
+}
+
+// GetCertificate implements the tls.Config GetClientCertificate() func
+func (cr *CertReloader) GetClientCertificate(req *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	cr.RLock()
+	defer cr.RUnlock()
+	return cr.certificate, nil
+}
+
+// TLSConfigApplyReloader patches a *tls.Config struct by setting the GetCertificate and GetClientCertificate
+// methods.
+func (cr *CertReloader) TLSConfigApplyReloader(cfg *tls.Config) {
+	cfg.GetCertificate = cr.GetCertificate
+	cfg.GetClientCertificate = cr.GetClientCertificate
+}
+
+func (cr *CertReloader) setCertificate() error {
+	log.Println("Loading TLS certificates...")
```

Comment:
I don't think libraries should be logging. It should at least be a toggle. Could also make the constructor take a logger.

P.S.  I really wish go authors had made log an interface.

Reply:
I agree and spent a better part of sunday thinking about it. I'll drop it for now.